packages/language-server/src/completion/completionUtil.ts

```diff
@@ -188,31 +188,72 @@ export function givenFieldAttributeParams(
     CompletionItemKind.Property,
   )
 
-  if (fieldAttribute === '@index') {
-    return items
-  }
+  return filterSortLengthBasedOnInput(
+    fieldAttribute,
+    previewFeatures,
+    datasourceProvider,
+    wordBeforePosition,
+    items,
+  )
+}
 
+export function filterSortLengthBasedOnInput(
+  attribute: '@@unique' | '@unique' | '@@id' | '@id' | '@@index' | '@@fulltext',
+  previewFeatures: PreviewFeatures[] | undefined,
+  datasourceProvider: string | undefined,
+  wordBeforePosition: string,
+  items: CompletionItem[],
+): CompletionItem[] {
   if (previewFeatures?.includes('extendedindexes')) {
-    // The sort argument is available for all databases on @unique
-    // The length argument is available on MySQL on @id, @unique
-    // Additionally, SQL Server also allows it on @id
-
     // Auto completion for Desc | Asc
     // includes because `@unique(sort: |)` means wordBeforePosition = '@unique(sort:'
     if (wordBeforePosition.includes('sort:')) {
       return sortAutoCompletionItems
     }
 
+    // The length argument is available on MySQL only on the
+    // @id, @@id, @unique, @@unique and @@index, @@fulltext fields.
+
+    // The sort argument is available for all databases on the
+    // @unique, @@unique and @@index fields.
+    // Additionally, SQL Server also allows it on @id and @@id.
+
+    // Which translates too
+    // - `length` argument for `@id`, `@@id`, `@unique`, `@@unique` and `@@index` (MySQL only)
+    // - Note that on the `@@` the argument is on available a field - not on the top level attribute
+    // - `sort` argument for `@unique`, `@@unique` and `@@index` (Additionally `@id` and `@@id` for SQL Server)
+
     if (datasourceProvider === 'mysql') {
-      return items
-    } else if (datasourceProvider === 'sqlserver') {
-      if (fieldAttribute === '@id') {
+      if (
+        ['@unique', '@@unique', '@@index', '@@fulltext'].includes(attribute)
+      ) {
         return items
       } else {
+        // filter sort out
+        return items.filter((arg) => arg.label !== 'sort')
+      }
+    } else if (datasourceProvider === 'sqlserver') {
+      if (
+        ['@unique', '@@unique', '@@index', '@id', '@@id'].includes(attribute)
+      ) {
+        // only filter length out
         return items.filter((arg) => arg.label !== 'length')
+      } else {
+        // filter length and sort out
+        return items.filter(
+          (arg) => arg.label !== 'length' && arg.label !== 'sort',
+        )
       }
     } else {
-      return items.filter((arg) => arg.label !== 'length')
+      if (['@unique', '@@unique', '@@index'].includes(attribute)) {
+        // only filter length out
+        return items.filter((arg) => arg.label !== 'length')
+      } else {
+        // filter length and sort out
+        return items.filter(
+          (arg) => arg.label !== 'length' && arg.label !== 'sort',
+        )
+      }
     }
```

Comment:
The rest looks correct, although the logic in that structure is hard to follow for me. Maybe we can later just unroll it and be explicit. This will lead to code duplication but allow us to check the rules for one provider in one place.

Reply:
Yes indeed...!

Reply:
I think we also need more tests